File: Average of Levels in Binary Tree - Leetcode 637/Average of Levels in Binary Tree - Leetcode 637.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = val;
 *     this.left = left === undefined ? null : left;
 *     this.right = right === undefined ? null : right;
 * }
 */

var averageOfLevels = function(root) {
    let avgs = [];
    if (root === null) return avgs;

    let q = [root];

    while (q.length > 0) {
        let sum = 0;
        let size = q.length;
        for (let i = 0; i < size; i++) {
            let node = q.shift();
            sum += node.val;

            if (node.left !== null) {
                q.push(node.left);
            }
            if (node.right !== null) {
                q.push(node.right);
            }
        }
        avgs.push(sum / size);
    }

    return avgs;
};

File: Average of Levels in Binary Tree - Leetcode 637/Average of Levels in Binary Tree - Leetcode 637.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        avgs = []
        q = deque()
        q.append(root)

        while q:
            summ = 0
            n = len(q)
            for _ in range(n):
                node = q.popleft()
                summ += node.val

                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

            avgs.append(summ / n)
        
        return avgs
        # Time: O(n)
        # Space: O(n)
